### PR TITLE
Add mqtt5 codec

### PR DESCRIPTION
**Pull request by alvin1221**

### CODE REVIEW

File: nanomq/sub_handler.c [hunk 1]
```diff
@@ -148,9 +94,9 @@ decode_sub_message(nano_work *work)
 
 		debug_msg("bpos+vpos: [%d]", bpos + vpos);
 		if (++bpos < remaining_len - vpos) {
-			if ((topic_node_t = nng_alloc(sizeof(topic_node))) ==
+			if ((topic_node_t = nng_zalloc(sizeof(topic_node))) ==
```
Comment (wanghaEMQ) [inline nanomq/sub_handler.c:97]: Why not nng_alloc?

Comment (alvin1221) [reply nanomq/sub_handler.c:97]:
```
void *
nni_alloc(size_t sz)
{
	return (sz > 0 ? malloc(sz) : NULL);
}

void *
nni_zalloc(size_t sz)
{
	return (sz > 0 ? calloc(1, sz) : NULL);
}
```

File: nanomq/unsub_handler.c [hunk 2]
```diff
@@ -153,7 +125,9 @@ encode_unsuback_message(nng_msg *msg, nano_work *work)
 
 #if SUPPORT_MQTT5_0
 	if (PROTOCOL_VERSION_v5 == proto_ver) {
-		nng_msg_append(msg, property_len, 1);
+		// nng_msg_append(msg, property_len, 1);
+		//TODO set property if necessary 
+		encode_properties(msg, NULL);
```
Comment (JaylinYu) [inline nanomq/unsub_handler.c:130]: this reduces performance

File: nanomq/apps/broker.c [hunk 3]
```diff
@@ -235,8 +236,8 @@ server_cb(void *arg)
 			nng_aio_set_msg(work->aio, work->msg);
 			nng_ctx_send(work->ctx, work->aio); // send connack
 
-			uint8_t *header = nng_msg_header(work->msg);
-			uint8_t  flag   = *(header + 3);
+			uint8_t *body = nng_msg_body(work->msg);
+			uint8_t  flag = *(body);
```
[the remark this answers is not in the document]
Comment (JaylinYu) [reply nanomq/apps/broker.c]: you can add the session present field here. but please get reason_code correctly.

File: nanomq/sub_handler.c [hunk 4]
```diff
@@ -322,14 +269,19 @@ sub_ctx_handle(nano_work *work)
 			tq = tq->next;
 		}
 		if (!topic_exist) {
-			// printf("Protocol Version: %d\n",
-			// old_ctx->proto_ver); printf("sub ID: %d\n",
-			// work->sub_pkt->sub_id.varint);
-			int          t = work->sub_pkt->sub_id.varint;
+			int            t = 0;
+			if (conn_param_get_protover(work->cparam) ==
```
Comment (JaylinYu) [inline nanomq/sub_handler.c:273]: racing condition

**Review by JaylinYu — CHANGES_REQUESTED** [review]
need more test

File: nanomq/pub_handler.c [hunk 5]
```diff
@@ -264,29 +285,36 @@ free_pub_packet(struct pub_packet_struct *pub_packet)
 {
```
Comment (JaylinYu) [inline nanomq/pub_handler.c:285]: what does this for?

File: nanomq/pub_handler.c [hunk 6]
```diff
@@ -322,17 +350,18 @@ bool
...
+	uint32_t arr_len    = 0;
+	int      append_res = 0;
+	uint8_t  proto      = 0;
+	uint32_t buf;
 
 	debug_msg("start encode message");
 
 	nng_msg_clear(dest_msg);
 	nng_msg_header_clear(dest_msg);
+	proto = conn_param_get_protover(work->cparam);
```
Comment (JaylinYu) [inline nanomq/pub_handler.c:363]: racing condition

**Review by JaylinYu — CHANGES_REQUESTED** [review]
racing condition everywhere

**Comment by JaylinYu** [comment]
I am closing this due to PR moved

**Pull request closed without merging**